Fix a bug that a root for an A/B test can be pruned during A/B test creation

dewei-zhu opened the pull request:
#### 7ab89ec17812da9b0187990f789cf32fad6cdff7
<pre>
Fix a bug that a root for an A/B test can be pruned during A/B test creation
<a href="https://bugs.webkit.org/show_bug.cgi?id=276265">https://bugs.webkit.org/show_bug.cgi?id=276265</a>
<a href="https://rdar.apple.com/122118306">rdar://122118306</a>

Reviewed by Ryosuke Niwa.

Fix a bug that a root uploaded during creation of an A/B test can be pruned because of other
root upload for current test or other tests by introducing &apos;uploadFileGracePeriodInHours&apos; config.
Fix a pruning issue that file can be double counted during pruning because a root can be referenced
by multiple commit sets.
Improve a pruning query to use &apos;commitset_requires_build&apos; isntead of &apos;commitset_patch_file IS NOT NULL&apos;
to determine if a root is built for a configuration. This increase the accuracy of the query because
a build request will be built if the other side of build request requires building.
Fix a bug that API response for &apos;/api/upload-root&apos; may contain stale file information if re-upload
happens.
Update file creation time while re-uploading a pruned file so that it will qualify the grace period for
a new uploaded file.

* Websites/perf.webkit.org/public/include/uploaded-file-helpers.php:
Add support for &apos;uploadFileGracePeriodInHours&apos;.
Fix a bug that &apos;/api/upload-root&apos; does not return latest information for a re-uploaded file.
Fix and improve the pruning query.
* Websites/perf.webkit.org/server-tests/privileged-api-upload-file-tests.js: Fix and add unit tests.
* Websites/perf.webkit.org/server-tests/resources/test-server.js: Add support to overwrite test config.
(TestServer.prototype.start):
(TestServer.prototype.writeTestConfig):
(TestServer.prototype.overwriteTestConfig):
(TestServer.prototype.inject):
(TestServer):

Canonical link: <a href="https://commits.webkit.org/280806@main">https://commits.webkit.org/280806@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/04b174f75837ef16b566d9cf5f74eefa4afb24cf

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57701 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37029 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10177 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61323 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8146 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44665 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8334 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46751 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5775 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59731 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34728 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49859 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27576 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31499 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7150 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/53451 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7422 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63004 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1615 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7497 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53971 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1621 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49870 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54084 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12761 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1366 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32858 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33944 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/35028 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33689 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Websites/perf.webkit.org/public/include/uploaded-file-helpers.php:
```diff
@@ -160,16 +163,17 @@ function delete_file($db, $file_row)
-    $user_filter = $remote_user ? 'AND file_author = $1' : 'AND file_author IS NULL';
-    $params = $remote_user ? array($remote_user) : array();
+    $file_filters = 'AND extract(epoch from file_created_at at time zone \'utc\') <= $1';
+    $file_filters .= ($remote_user ? ' AND file_author = $2' : ' AND file_author IS NULL');
```
This parenthesis seems unnecessary?

dewei-zhu replied:
No, will remove it.

rniwa commented on Websites/perf.webkit.org/public/include/uploaded-file-helpers.php:
```diff
@@ -160,16 +163,17 @@ function delete_file($db, $file_row)
-    // 1. Delete old build products created for a patch not associated with any pending or in-progress builds.
-    $build_product_query = $db->query("SELECT file_id, file_extension, file_size FROM uploaded_files, commit_set_items
-        WHERE file_id = commitset_root_file AND commitset_patch_file IS NOT NULL AND file_deleted_at IS NULL
+    // 1. Delete old build products created not associated with any pending or in-progress builds.
```
*Delete old build products not associated* or *Delete old build products created for any pending...*

rniwa commented on Websites/perf.webkit.org/public/include/uploaded-file-helpers.php:
```diff
@@ -160,16 +163,17 @@ function delete_file($db, $file_row)
-    $build_product_query = $db->query("SELECT file_id, file_extension, file_size FROM uploaded_files, commit_set_items
-        WHERE file_id = commitset_root_file AND commitset_patch_file IS NOT NULL AND file_deleted_at IS NULL
+    // 1. Delete old build products created not associated with any pending or in-progress builds.
+    $build_product_query = $db->query("SELECT DISTINCT file_id, file_extension, file_size, file_created_at FROM uploaded_files, commit_set_items
```
Why do we need DISTINCT here?

rniwa replied:
Oh, this is the fix for double counting. It would be good to add a per-function / file description in the commit message.

rniwa commented on Websites/perf.webkit.org/public/include/uploaded-file-helpers.php:
```diff
@@ -92,9 +92,11 @@ function upload_file_in_transaction($db, $input_file, $remote_user, $additional_
     $new_file_size = $input_file['size'];
     if (config('uploadUserQuotaInMB') * MEGABYTES - query_file_usage_for_user($db, $remote_user) < $new_file_size
         || config('uploadTotalQuotaInMB') * MEGABYTES - query_total_file_usage($db) < $new_file_size) {
+        $max_database_time_ahead_php_time_in_seconds = 60;
```
What is this?

dewei-zhu committed:
Fix a bug that a root for an A/B test can be pruned during A/B test creation
https://bugs.webkit.org/show_bug.cgi?id=276265
rdar://122118306

Reviewed by Ryosuke Niwa.

Fix a bug that a root uploaded during creation of an A/B test can be pruned because of other
root upload for current test or other tests by introducing 'uploadFileGracePeriodInHours' config.
Fix a pruning issue that file can be double counted during pruning because a root can be referenced
by multiple commit sets.
Improve a pruning query to use 'commitset_requires_build' isntead of 'commitset_patch_file IS NOT NULL'
to determine if a root is built for a configuration. This increase the accuracy of the query because
a build request will be built if the other side of build request requires building.
Fix a bug that API response for '/api/upload-root' may contain stale file information if re-upload
happens.
Update file creation time while re-uploading a pruned file so that it will qualify the grace period for
a new uploaded file.

* Websites/perf.webkit.org/public/include/uploaded-file-helpers.php:
Add support for 'uploadFileGracePeriodInHours'.
Fix a bug that '/api/upload-root' does not return latest information for a re-uploaded file.
Fix and improve the pruning query.
* Websites/perf.webkit.org/server-tests/privileged-api-upload-file-tests.js: Fix and add unit tests.
* Websites/perf.webkit.org/server-tests/resources/test-server.js: Add support to overwrite test config.
(TestServer.prototype.start):
(TestServer.prototype.writeTestConfig):
(TestServer.prototype.overwriteTestConfig):
(TestServer.prototype.inject):
(TestServer):

Canonical link: https://commits.webkit.org/280806@main
